From 0ea3e17409a899d627e222b768180e8fcd2b3980 Mon Sep 17 00:00:00 2001
From: Richard Nelson <wabz@pidgin.im>
Date: Mon, 6 Apr 2009 23:05:02 +0000
Subject: [PATCH] Set _MESSAGE_{RECV,SEND} properly when receiving chat
 messages. Thanks Sadrul!

--HG--
extra : convert_revision : eed0598058a6e7cc5da7ad99f06bfe136291d2f9
---
 libpurple/server.c |   14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/libpurple/server.c b/libpurple/server.c
index 5da41a9..d3104b8 100644
--- a/libpurple/server.c
+++ b/libpurple/server.c
@@ -925,12 +925,6 @@ void serv_got_chat_in(PurpleConnection *g, int id, const char *who,
 	g_return_if_fail(who != NULL);
 	g_return_if_fail(message != NULL);
 
-	/*
-	 * XXX: Should we be setting this here, or relying on prpls to set it?
-	 */
-	if (g_strcmp0(purple_account_get_username(g->account), who))
-		flags |= PURPLE_MESSAGE_RECV;
-
 	for (bcs = g->buddy_chats; bcs != NULL; bcs = bcs->next) {
 		conv = (PurpleConversation *)bcs->data;
 
@@ -945,6 +939,14 @@ void serv_got_chat_in(PurpleConnection *g, int id, const char *who,
 	if (!conv)
 		return;
 
+	/* Did I send the message? */
+	if (purple_strequal(purple_conv_chat_get_nick(chat), who)) {
+		flags |= PURPLE_MESSAGE_SEND;
+		flags &= ~PURPLE_MESSAGE_RECV; /* Just in case some prpl sets it! */
+	} else {
+		flags |= PURPLE_MESSAGE_RECV;
+	}
+
 	/*
 	 * Make copies of the message and the sender in case plugins want
 	 * to free these strings and replace them with a modifed version.
-- 
1.7.10.4

